OpenBLAS tests fail `potrf:smoketest_trivial` with LLVM 20.1.8 on aarch64

Heya 👋 

I'm currently looking into building software packages for a potential LLVM toolchain in [EasyBuild](https://easybuild.io/).

While trying to build OpenBLAS on a Grace Hopper system, I noticed that OpenBLAS fails to run a single test in this configuration:

```console
$ make -j 16 shared  BINARY='64'  CC='clang'  FC='flang'  MAKE_NB_JOBS='-1'  USE_OPENMP='1'  USE_THREAD='1'  CFLAGS='-O2 -Wl,--undefined-version -Wno-unused-command-line-argument -Wno-error=int-conversion'
$ OMP_NUM_THREADS=16 OPENBLAS_NUM_THREADS=16   make tests  BINARY='64'  CC='clang'  FC='flang'  MAKE_NB_JOBS='-1'  USE_OPENMP='1'  USE_THREAD='1'
```

With any number of threads above 1, we see this failure (I've added the `(%d)` at the end to get the actual `info` value):

```
TEST 124/126 potrf:smoketest_trivial [FAIL]
  ERR: test_potrs.c:463  info != 0 (2)
```

With a single thread, this part passes, but we fail later on:

```
TEST 124/126 potrf:smoketest_trivial [FAIL]
  ERR: test_potrs.c:537  U s(1,1) difference: 0.0449998
```

Given that `flang` is not entirely mature, I wouldn't be surprised if there's still a compiler bug somewhere.
However, I failed to find a trace of this being reported in LLVM or here.

The same configuration worked just fine on Intel Alderlake and AMD Zen4, so I'm guessing that this is an aarch64 specific issue.


----

System info: Rocky Linux 9.6, NVIDIA Grace Hopper GH200, 480 GB RAM

Used software:

```console
$ module list
Currently Loaded Modules:
  1) GCCcore/14.3.0   3) binutils/2.44   5) XZ/5.8.1         7) ncurses/6.5   9) Z3/4.15.1  11) lz4/1.10.0  13) LLVM/20.1.8            15) make/4.4.1   17) libreadline/8.2   19) Tcl/9.0.1      21) OpenSSL/3
  2) zlib/1.3.1       4) libffi/3.5.1    6) libxml2/2.14.3   8) GMP/6.3.0    10) gzip/1.14  12) zstd/1.5.7  14) llvm-compilers/20.1.8  16) bzip2/1.0.8  18) libtommath/1.3.0  20) SQLite/3.50.1  22) Python/3.13.5
```

----

If any more information is needed, I'm happy to help out. Unforunately, I don't have much of an idea where this issue could be coming from...

Testing from source:

**CMake:**

```console
$ cmake -S .. -B . -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_Fortran_COMPILER=flang
$ make -j12
$ ctest --output-on-failure
[...]
112/116 Test #112: xccblat2 ....................................................................................   Passed    0.70 sec
        Start 113: xccblat3
113/116 Test #113: xccblat3 ....................................................................................   Passed    0.15 sec
        Start 114: xzcblat1
114/116 Test #114: xzcblat1 ....................................................................................   Passed    0.00 sec
        Start 115: xzcblat2
115/116 Test #115: xzcblat2 ....................................................................................   Passed    0.70 sec
        Start 116: xzcblat3
116/116 Test #116: xzcblat3 ....................................................................................   Passed    1.18 sec

100% tests passed, 0 tests failed out of 116

Total Test time (real) = 123.32 sec
```

- [openblas-cmake.log](https://github.com/user-attachments/files/23419104/openblas-cmake.log)

**make:**

```console
$ make -j 16 shared  BINARY='64'  CC='clang'  FC='flang'  MAKE_NB_JOBS='-1'  USE_OPENMP='1'  USE_THREAD='1'
$ make tests  BINARY='64'  CC='clang'  FC='flang'  MAKE_NB_JOBS='-1'  USE_OPENMP='1'  USE_THREAD='1'
[...]
TEST 117/126 zgemv:0_nan_inf [OK]
TEST 118/126 zgemv:0_nan_inf_incy_2 [OK]
TEST 119/126 zgemv:0_2_nan_1_inf_1 [OK]
TEST 120/126 zgemv:0_2_nan_1_inf_1_incy_2 [OK]
TEST 121/126 zgemv:2_0_nan_1_inf_1 [OK]
TEST 122/126 zgemv:2_0_nan_1_inf_1_incy_2 [OK]
TEST 123/126 potrf:bug_695 [OK]
TEST 124/126 potrf:smoketest_trivial [FAIL]
  ERR: test_potrs.c:463  info != 0
TEST 125/126 kernel_regress:skx_avx [OK]
TEST 126/126 fork:safety_after_fork_in_parent [OK]
RESULTS: 126 tests (125 ok, 1 failed, 0 skipped) ran in 7 ms
make[1]: *** [Makefile:86: run_test] Error 1
make[1]: Leaving directory '/p/project1/cswmanage/reuter1/OpenBLAS/utest'
make: *** [Makefile:177: tests] Error 2
```

Just by quickly glancing over the CMake results, they look better I guess?

Well, looking at it more closely, one can see 
```
Fortran STOP
IEEE arithmetic exceptions signaled: INEXACT
``` 
in quite a few tests. That's probably not ideal.

--- a/.github/workflows/dynamic_arch.yml
+++ b/.github/workflows/dynamic_arch.yml
@@ -1,6 +1,6 @@
 name: continuous build
 
-on: [push, pull_request]
+on: [push, pull_request, workflow_dispatch]
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
@@ -11,17 +11,24 @@ permissions:
 
 jobs:
   build:
-    if: "github.repository == 'OpenMathLib/OpenBLAS'"
+    if: "github.repository == 'OpenMathLib/OpenBLAS' || github.event_name == 'workflow_dispatch'"
     runs-on: ${{ matrix.os }}
 
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-latest, macos-latest]
+        os: [ubuntu-latest, macos-latest, ubuntu-24.04-arm]
+        cc: [gcc, clang, clang-21]
         fortran: [gfortran, flang]
         build: [cmake, make]
         exclude:
           - os: macos-latest
+            cc: gcc
+          - os: macos-latest
+            cc: clang-21
+          - os: macos-latest
+            fortran: flang
+          - os: ubuntu-24.04-arm
             fortran: flang
 
     steps:
@@ -42,10 +49,23 @@ jobs:
       - name: Install Dependencies
         run: |
           if [ "$RUNNER_OS" == "Linux" ]; then
+            cat << EOF | sudo tee -a /etc/apt/apt.conf.d/01norecommend
+          APT::Install-Recommends "0";
+          APT::Install-Suggests "0";
+          EOF
             sudo apt-get update
-            sudo apt-get install -y gfortran cmake ccache
-            wget http://security.ubuntu.com/ubuntu/pool/universe/n/ncurses/libtinfo5_6.3-2ubuntu0.1_amd64.deb
-            sudo apt install ./libtinfo5_6.3-2ubuntu0.1_amd64.deb
+            sudo apt-get install -y ccache
+            if [ "${{ matrix.cc }}" == "clang-21" ]; then
+                wget https://apt.llvm.org/llvm.sh
+                chmod +x llvm.sh
+                sudo ./llvm.sh 21
+            fi
+            if [ "${{ matrix.fortran }}" == "flang" ]; then
+              wget http://security.ubuntu.com/ubuntu/pool/universe/n/ncurses/libtinfo5_6.3-2ubuntu0.1_amd64.deb
+              sudo apt install ./libtinfo5_6.3-2ubuntu0.1_amd64.deb
+            else
+              sudo apt-get install -y ${{ matrix.fortran }}
+            fi
           elif [ "$RUNNER_OS" == "macOS" ]; then
             # It looks like "gfortran" isn't working correctly unless "gcc" is re-installed.
             brew reinstall gcc
@@ -64,12 +84,12 @@ jobs:
           # GNU make and cmake call the compilers differently. It looks like
           # that causes the cache to mismatch. Keep the ccache for both build
           # tools separate to avoid polluting each other.
-          key: ccache-${{ runner.os }}-${{ matrix.build }}-${{ matrix.fortran }}-${{ github.ref }}-${{ github.sha }}
+          key: ccache-${{ runner.os }}-${{ runner.arch }}-${{ matrix.build }}-${{ matrix.cc }}-${{ matrix.fortran }}-${{ github.ref }}-${{ github.sha }}
           # Restore a matching ccache cache entry. Prefer same branch and same Fortran compiler.
           restore-keys: |
-            ccache-${{ runner.os }}-${{ matrix.build }}-${{ matrix.fortran }}-${{ github.ref }}
-            ccache-${{ runner.os }}-${{ matrix.build }}-${{ matrix.fortran }}
-            ccache-${{ runner.os }}-${{ matrix.build }}
+            ccache-${{ runner.os }}-${{ runner.arch }}-${{ matrix.build }}-${{ matrix.cc }}-${{ matrix.fortran }}-${{ github.ref }}
+            ccache-${{ runner.os }}-${{ runner.arch }}-${{ matrix.build }}-${{ matrix.cc }}-${{ matrix.fortran }}
+            ccache-${{ runner.os }}-${{ runner.arch }}-${{ matrix.build }}-${{ matrix.cc }}
 
       - name: Configure ccache
         run: |
@@ -89,7 +109,7 @@ jobs:
           echo "max_size = 300M" > ~/.ccache/ccache.conf
           echo "compression = true" >> ~/.ccache/ccache.conf
           ccache -s
-          
+
       - name: Add gfortran runtime to link path
         if: matrix.build == 'make' && runner.os == 'macOS'
         run: |
@@ -110,7 +130,7 @@ jobs:
           fi
           case "${{ matrix.build }}" in
             "make")
-              make -j$(nproc) DYNAMIC_ARCH=1 USE_OPENMP=0 FC="ccache ${{ matrix.fortran }}"
+              make -j$(nproc) DYNAMIC_ARCH=1 USE_OPENMP=0 CC="ccache ${{ matrix.cc }}" FC="ccache ${{ matrix.fortran }}"
               ;;
             "cmake")
               mkdir build && cd build
@@ -119,6 +139,7 @@ jobs:
                     -DBUILD_WITHOUT_LAPACK=0 \
                     -DCMAKE_VERBOSE_MAKEFILE=ON \
                     -DCMAKE_BUILD_TYPE=Release \
+                    -DCMAKE_C_COMPILER=${{ matrix.cc }} \
                     -DCMAKE_Fortran_COMPILER=${{ matrix.fortran }} \
                     -DCMAKE_C_COMPILER_LAUNCHER=ccache \
                     -DCMAKE_Fortran_COMPILER_LAUNCHER=ccache \
@@ -142,13 +163,13 @@ jobs:
             "make")
               MAKE_FLAGS='DYNAMIC_ARCH=1 USE_OPENMP=0'
               echo "::group::Tests in 'test' directory"
-              make -C test $MAKE_FLAGS FC="ccache ${{ matrix.fortran }}"
+              make -C test $MAKE_FLAGS CC="ccache ${{ matrix.cc }}" FC="ccache ${{ matrix.fortran }}"
               echo "::endgroup::"
               echo "::group::Tests in 'ctest' directory"
-              make -C ctest $MAKE_FLAGS FC="ccache ${{ matrix.fortran }}"
+              make -C ctest $MAKE_FLAGS CC="ccache ${{ matrix.cc }}" FC="ccache ${{ matrix.fortran }}"
               echo "::endgroup::"
               echo "::group::Tests in 'utest' directory"
-              make -C utest $MAKE_FLAGS FC="ccache ${{ matrix.fortran }}"
+              make -C utest $MAKE_FLAGS CC="ccache ${{ matrix.cc }}" FC="ccache ${{ matrix.fortran }}"
               echo "::endgroup::"
               ;;
             "cmake")
@@ -372,15 +393,15 @@ jobs:
     steps:
       - name: Checkout repository
         uses: actions/checkout@v3
-        
+
       - name: Install Dependencies
         run: |
           sudo apt-get update
           sudo apt-get install -y gcc gfortran make
-          
+
       - name: Build OpenBLAS
         run: |
-          make -j${nproc} 
+          make -j${nproc}
           make -j${nproc} lapack-test
-    
-          
+
+

--- a/kernel/arm64/dot_kernel_sve.c
+++ b/kernel/arm64/dot_kernel_sve.c
@@ -48,7 +48,7 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 "        add     x12, %[X_], x9, lsl #"SHIFT"               \n" \
 "        add     x13, %[Y_], x9, lsl #"SHIFT"               \n"
 #define TAIL_WHILE                                        \
-"        whilelo p1."DTYPE", x8, x0                         \n"
+"        whilelo p1."DTYPE", x8, %[N_]                         \n"
 #define UPDATE(pg, x,y,out)                               \
 "        ld1"WIDTH"    { z2."DTYPE" }, "pg"/z, ["x", x8, lsl #"SHIFT"]  \n" \
 "        ld1"WIDTH"    { z3."DTYPE" }, "pg"/z, ["y", x8, lsl #"SHIFT"]  \n" \
@@ -66,7 +66,7 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 "        movi    d1, #0x0                             \n" \
         SETUP_TRUE                                        \
 "        neg     x10, x9, lsl #1                      \n" \
-"        ands    x11, x10, x0                         \n" \
+"        ands    x11, x10, %[N_]                      \n" \
 "        b.eq    2f // skip_2x                        \n" \
         OFFSET_INPUTS                                     \
 "1: // vector_2x                                      \n" \
@@ -78,7 +78,7 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
         SUM_VECTOR("1") \
 "2: // skip_2x                                        \n" \
 "        neg     x10, x9                              \n" \
-"        and     x10, x10, x0                         \n" \
+"        and     x10, x10, %[N_]                      \n" \
 "        cmp     x8, x10                              \n" \
 "        b.hs    4f // tail                           \n" \
 "3: // vector_1x                                      \n" \
@@ -87,7 +87,7 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 "        cmp     x8, x10                              \n" \
 "        b.lo    3b // vector_1x                      \n" \
 "4: // tail                                           \n" \
-"        cmp     x10, x0                              \n" \
+"        cmp     x10, %[N_]                           \n" \
 "        b.eq    5f // end                            \n" \
         TAIL_WHILE                                        \
         UPDATE("p1", "%[X_]", "%[Y_]", "z0")              \
@@ -110,9 +110,9 @@ dot_kernel_sve(BLASLONG n, FLOAT* x, FLOAT* y)
         [Y_] "r" (y)
       : "cc",
         "memory",
-        "x0", "x1", "x2", "x3", "x4", "x5", "x6", "x7",
-        "x8", "x9", "x10", "x11", "x12", "x13", "d1", 
-        "z0", "z1"
+        "x8", "x9", "x10", "x11", "x12", "x13",
+        "v0", "v1",
+        "z0", "z1", "z2", "z3"
   );
 
   return ret;

